File: Board.java
import java.util.*;
import java.io.*;

public class Board {
    // board private vars
	private char[][] board;
    private int xdim, ydim, size, fitness, numMines;
    private Random rgen;

    // create a board with the given dimensions and number of mines
    public Board(int xdim, int ydim, int numMines) {
        this.xdim = xdim;
        this.ydim = ydim;
        this.numMines = numMines;
		this.size = xdim*ydim;
		this.fitness = this.size - numMines;
		this.rgen = new Random();
        this.board = new char[xdim][ydim];
        // initially fill the grid with zeros
        for (int i=0; i<xdim; i++) {
            for (int j=0; j<ydim; j++) {
                board[i][j] = ' ';
			}
		}
    }
	
	public int getFitness() {
		return this.fitness;
	}
	
	public void placeMines() {
		int mineLocation, i;
		for (i=0; i<numMines; i++){
			mineLocation = rgen.nextInt(size);
			
			// If I reimplement this in C, do something smart like only get the address
			// of the cell once and use it twice.
			if (board[mineLocation/xdim][mineLocation%ydim] == ' ') {
				board[mineLocation/xdim][mineLocation%ydim] = '*';
			} else {
				i--;
			}
		}
	}
	
	public void endGame() {
		System.out.println("😵 You lost. Your Fitness was: " + fitness);
	}
	
	public void calculateAdjacentMines(int x, int y) {
		int count = 0;
		for (int i = x-1; i < x+2; i++) {
			for (int j = y-1; j < y+2; j++) {
				if (i == x && j == y) {
					continue;
				} else {
					try {
						if (board[i][j] == '*') {
							count++;
						}
					}
					catch (IndexOutOfBoundsException e) {
						continue;
					}
				}
			}
		}

		board[x][y] = Integer.toString(count).charAt(0);
		if (count == 0) {
			handleAdjacent(x,y);
		}
	}
	
	public void handleAdjacent(int x, int y) {
		for (int i = x-1; i < x+2; i++) {
			for (int j = y-1; j < y+2; j++) {
				try {
					if (i == x && j == y) {
						continue;
					} else if (board[i][j] == ' ') {
							click(i,j);
					}
				} catch (IndexOutOfBoundsException e) {
					continue;
				}
			}
		}
	}
	
	public int click(int x, int y) {
		if (board[x][y] == '*') {
			board[x][y] = '!';
			endGame();
			return 1;
		}
		else if (board[x][y] == ' ')
			{
			calculateAdjacentMines(x,y);
			fitness--;
		} else {
			System.out.println("You've already discovered this cell");
		}
		return 0;
	}
	
    public void print() {
        print(System.out);
    }

    // print the current state of the grid, showing blocks and the dozer 
    // pointing in the correct direction
    public void print(PrintStream os) {
        for (int y=0; y<ydim; y++) {
            os.print("|");
			String line = "\n";
            for (int x=0; x<xdim; x++) {
                char out = board[x][y];
				line+="  - ";
				if (out == '*') {
					out = ' ';
				}
                os.print(" " + out+" |");
            }
            os.println (line);
        }
        os.println();
		os.println("Fitness = " + fitness);
    }

	public static void main(String[] args) {
		Board board = new Board(10,10,10);
		board.placeMines();
		board.print();
		Scanner reader = new Scanner(System.in);
		int x,y;
		int loop = 0;
		while (loop == 0) {
			System.out.println("Enter the coordinates");
			x = reader.nextInt();
			y = reader.nextInt();
			loop = board.click(x,y);
			board.print();
			if (board.getFitness() == 0) {
				System.out.println(" 😎 You Win!");
				loop = 1;
			}
		}
	}
}

File: board.py
'''
Michael Wheatman
2015
Minesweeper Game for NEAT Algorithm tests
'''


import random
import sys
 
class Board:
	def __init__(self, size, numMines):
		self.board = [[' ' for x in range(size)] for x in range(size)]
		self.size = size
		self.fitness = size**2 - numMines
		self.numMines = numMines

	def getBoard(self):
		return self.board
	
	def getSize(self):
		return self.size
	
	def getFitness(self):
		return self.fitness

	def placeMines(self):
		for i in range(self.numMines):
			mineLocation = random.randrange(self.size**2)
			if self.board[mineLocation/self.size][mineLocation%self.size] == ' ':
				self.board[mineLocation/self.size][mineLocation%self.size] = '*'
			else:
				i-=1

	def endGame(self):
		print "You lost. Your Fitness was: " + self.fitness

	def calculateAdjacentMines(self, x, y):
		count = 0
		for i,j in [[x-1, y-1], [x-1,y] , [x-1, y+1],
					[x, y-1], [x, y+1],
					[x+1, y-1], [x+1,y] , [x+1, y+1]]:
			if i>-1 and j >-1 and i<10 and j < 10:
				if self.board[i][j] == '*':
					count+=1
		if count == 0:
			self.board[x][y] = count
			self.handleAdjacent(x,y)
		else:
			self.board[x][y] = count

	def handleAdjacent(self, x, y):
		for i,j in [[x-1, y-1], [x-1,y] , [x-1, y+1],
					[x, y-1], [x, y+1],
					[x+1, y-1], [x+1,y] , [x+1, y+1]]:
			if i>-1 and j >-1 and i<10 and j < 10:
				if self.board[i][j] == ' ':
						self.click(i,j)

	def click(self, x, y):
		if self.board[x][y] == '*':
			self.board[x][y] = '!'
			self.endGame()
			return 1
		elif self.board[x][y] == ' ':
			self.fitness-=1
			self.calculateAdjacentMines(x,y)
			
		else:
			print "You've already discovered this cell"
		return 0
	
	def initialClick(self,x,y):
		self.board[x][y] = '_'
		self.placeMines()
		self.fitness-=1
		self.calculateAdjacentMines(x,y)

	def printBoard(self):
		header = "y \\ x\t"
		for i in range(self.size):
			header+="  "+`i`+ " "
		print header
		for y in range(self.size):
			sys.stdout.write(`y`+"\t"+ "|")
			line = "\n\t"
			for x in range(self.size):
				out = str(self.board[x][y])
				line+="  - "
				if out == '*':
					out = '*'
				sys.stdout.write(" "+out+" |")
			print line
		print self.fitness

def main():
	board = Board(10,3)
	loop = 0
	started = 0
	while loop == 0:
		board.printBoard()
		coords = raw_input('Enter the coordinates: ').split()
		x = int(coords[0])
		y = int(coords[1])
		if started == 0:
			board.initialClick(x,y)
		else:
			loop = board.click(x,y);
		if board.getFitness() == 0:
			print "You Win!"
			loop = 1

if __name__ == "__main__":
	main()
